File: samples/nodejs/entra-agent-id-sidecar/src/callerCard.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import { Attachment } from '@microsoft/agents-activity'
import { CardFactory, TurnContext } from '@microsoft/agents-hosting'

interface CardFact {
  title: string;
  value: string;
}

export function createCallerCard (context: TurnContext): Attachment {
  const claims = context.identity as Record<string, unknown> | undefined
  const recipient = context.activity.recipient
  const authenticated = Boolean(claims && Object.keys(claims).length > 0)
  const facts: CardFact[] = [
    fact('Claims available', authenticated ? 'Yes' : 'No'),
    fact('Authenticated', authenticated ? 'Yes' : 'No'),
    fact('Authentication type', authenticated ? 'JWT bearer' : undefined),
    fact('Issuer', claim(claims, 'iss')),
    fact('Audience', claim(claims, 'aud')),
    fact('Tenant ID', claim(claims, 'tid', 'http://schemas.microsoft.com/identity/claims/tenantid')),
    fact('Subject', claim(claims, 'sub', 'http://schemas.xmlsoap.org/ws/2005/05/identity/claims/nameidentifier')),
    fact('Caller app ID', claim(claims, 'azp', 'appid')),
    fact('Token version', claim(claims, 'ver')),
    fact('Issued', epochClaim(claims, 'iat')),
    fact('Expires', epochClaim(claims, 'exp')),
    fact('Recipient role', recipient?.role),
    fact('Agent identity', recipient?.agenticAppId),
    fact('Agent user', recipient?.agenticUserId),
    fact('Activity tenant', recipient?.tenantId)
  ]

  return CardFactory.adaptiveCard({
    type: 'AdaptiveCard',
    version: '1.5',
    body: [
      {
        type: 'TextBlock',
        size: 'Medium',
        weight: 'Bolder',
        text: 'Inbound agentic caller'
      },
      {
        type: 'TextBlock',
        wrap: true,
        text: 'Selected token claims and activity routing values. The raw token is never displayed.'
      },
      {
        type: 'FactSet',
        facts
      }
    ]
  })
}

function fact (title: string, value: unknown): CardFact {
  const text = Array.isArray(value) ? value.join(', ') : String(value ?? '').trim()
  return { title, value: text || 'Unavailable' }
}

function claim (claims: Record<string, unknown> | undefined, ...names: string[]): unknown {
  for (const name of names) {
    if (claims?.[name] !== undefined) {
      return claims[name]
    }
  }

  return undefined
}

function epochClaim (claims: Record<string, unknown> | undefined, name: string): string | undefined {
  const value = claim(claims, name)
  const seconds = typeof value === 'number' ? value : Number(value)
  return Number.isFinite(seconds) ? new Date(seconds * 1000).toISOString() : undefined
}

File: samples/python/entra-agent-id-sidecar/src/caller_card.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License.

from datetime import datetime, timezone
from typing import Any

from microsoft_agents.hosting.core import CardFactory, TurnContext


def create_caller_card(context: TurnContext):
    identity = context.identity
    claims = identity.claims if identity else {}
    recipient = context.activity.recipient

    facts = [
        _fact("Claims available", "Yes" if claims else "No"),
        _fact("Authenticated", "Yes" if identity and identity.is_authenticated else "No"),
        _fact("Authentication type", identity.authentication_type if identity else None),
        _fact("Issuer", _claim(claims, "iss")),
        _fact("Audience", _claim(claims, "aud")),
        _fact(
            "Tenant ID",
            _claim(claims, "tid", "http://schemas.microsoft.com/identity/claims/tenantid"),
        ),
        _fact(
            "Subject",
            _claim(
                claims,
                "sub",
                "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/nameidentifier",
            ),
        ),
        _fact("Caller app ID", _claim(claims, "azp", "appid")),
        _fact("Token version", _claim(claims, "ver")),
        _fact("Issued", _epoch_claim(claims, "iat")),
        _fact("Expires", _epoch_claim(claims, "exp")),
        _fact("Recipient role", recipient.role if recipient else None),
        _fact("Agent identity", recipient.agentic_app_id if recipient else None),
        _fact("Agent user", recipient.agentic_user_id if recipient else None),
        _fact("Activity tenant", recipient.tenant_id if recipient else None),
    ]

    return CardFactory.adaptive_card(
        {
            "type": "AdaptiveCard",
            "version": "1.5",
            "body": [
                {
                    "type": "TextBlock",
                    "size": "Medium",
                    "weight": "Bolder",
                    "text": "Inbound agentic caller",
                },
                {
                    "type": "TextBlock",
                    "wrap": True,
                    "text": (
                        "Selected token claims and activity routing values. "
                        "The raw token is never displayed."
                    ),
                },
                {"type": "FactSet", "facts": facts},
            ],
        }
    )


def _fact(title: str, value: Any) -> dict[str, str]:
    if isinstance(value, list):
        value = ", ".join(str(item) for item in value)
    text = str(value).strip() if value is not None else ""
    return {"title": title, "value": text or "Unavailable"}


def _claim(claims: dict[str, Any], *names: str):
    return next((claims[name] for name in names if name in claims), None)


def _epoch_claim(claims: dict[str, Any], name: str):
    value = _claim(claims, name)
    try:
        return datetime.fromtimestamp(int(value), timezone.utc).isoformat()
    except (TypeError, ValueError, OverflowError):
        return value
